### PR TITLE
Blocked k-NN index for replication

### PR DESCRIPTION
**Pull request by saikaranam-amazon**
Signed-off-by: Sai Kumar <karanas@amazon.com>

### Description
Blocked k-NN index for replication

UTs - https://github.com/opensearch-project/cross-cluster-replication/issues/159
 
### Issues Resolved
```
amazon.opendistroforelasticsearch.knn.plugin.KNNEngineFactory]]], allocation_status[fetching_shard_data]], expected_shard_size[208], message [failed to create index], failure [IllegalStateException[multiple engine factories provided for [leader-test-knn/XPgNHnwCT_-Yj3QcuxqRyQ]: [com.amazon.elasticsearch.replication.ReplicationPlugin$getEngineFactory$1],[com.amazon.opendistroforelasticsearch.knn.plugin.KNNEngineFactory]]], markAsStale [true]]
java.lang.IllegalStateException: multiple engine factories provided for [leader-test-knn/XPgNHnwCT_-Yj3QcuxqRyQ]: [com.amazon.elasticsearch.replication.ReplicationPlugin$getEngineFactory$1],[com.amazon.opendistroforelasticsearch.knn.plugin.KNNEngineFactory]
        at org.elasticsearch.indices.IndicesService.getEngineFactory(IndicesService.java:705)
        at org.elasticsearch.indices.IndicesService.createIndexService(IndicesService.java:646)
        at org.elasticsearch.indices.IndicesService.createIndex(IndicesService.java:558)
        at org.elasticsearch.indices.IndicesService.createIndex(IndicesService.java:177)
        at org.elasticsearch.indices.cluster.IndicesClusterStateService.createIndices(In

```
 
### Check List
- [x] New functionality includes testing.
  - [ ] All tests pass
- [x] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by ankitkala — APPROVED** [review]
LGTM

File: src/main/kotlin/org/opensearch/replication/action/resume/TransportResumeIndexReplicationAction.kt [hunk 1]
```diff
@@ -97,7 +99,16 @@ class TransportResumeIndexReplicationAction @Inject constructor(transportService
...
+
+                val leaderSettings = settingsResponse.indexToSettings.get(params.leaderIndex.name) ?: throw IndexNotFoundException(params.leaderIndex.name)
+
+                // k-NN Setting is a static setting. In case the setting is changed at the leader index before resume,
+                // block the resume.
+                if(leaderSettings.getAsBoolean(KNN_INDEX_SETTING, false)) {
+                    throw IllegalStateException("Index setting (index.knn) changed. Cannot resume k-NN index - ${params.leaderIndex.name}")
+                }
+
+                ValidationUtil.validateAnalyzerSettings(environment, leaderSettings, replMetdata.settings)
```
Comment (krishna-ggk) [inline src/main/kotlin/org/opensearch/replication/action/resume/TransportResumeIndexReplicationAction.kt:111]: Minor messaging feedback - "Cannot resume replication for k-NN enabled index {index}." ?

File: src/main/kotlin/org/opensearch/replication/action/index/TransportReplicateIndexAction.kt [hunk 2]
```diff
@@ -89,6 +90,13 @@ class TransportReplicateIndexAction @Inject constructor(transportService: Transp
                 if (!leaderSettings.getAsBoolean(IndexSettings.INDEX_SOFT_DELETES_SETTING.key, true)) {
                     throw IllegalArgumentException("Cannot Replicate an index where the setting ${IndexSettings.INDEX_SOFT_DELETES_SETTING.key} is disabled")
                 }
+
+                // For k-NN indices, k-NN loads its own engine and this conflicts with the replication follower engine
+                // Blocking k-NN indices for replication
+                if(leaderSettings.getAsBoolean(KNN_INDEX_SETTING, false)) {
+                    throw IllegalArgumentException("Cannot Replicate k-NN index - ${request.leaderIndex}")
```
Comment (krishna-ggk) [inline src/main/kotlin/org/opensearch/replication/action/index/TransportReplicateIndexAction.kt:97]: Minor - lower case 'R' in Replicate

**Review by krishna-ggk — APPROVED** [review]

File: src/main/kotlin/org/opensearch/replication/action/resume/TransportResumeIndexReplicationAction.kt (hunk 1 above)
Comment (saikaranam-amazon) [reply src/main/kotlin/org/opensearch/replication/action/resume/TransportResumeIndexReplicationAction.kt]: Done

File: src/main/kotlin/org/opensearch/replication/action/index/TransportReplicateIndexAction.kt (hunk 2 above)
Comment (saikaranam-amazon) [reply src/main/kotlin/org/opensearch/replication/action/index/TransportReplicateIndexAction.kt]: Done

**Review by krishna-ggk — APPROVED** [review]

**Review by rushiagr — APPROVED** [review]

**Pull request merged**